### PR TITLE
agents: harness-review hardening of the subagent system (definition-of-done pass)

### PR DESCRIPTION
**Pull request by agjs**
Full adversarial **harness-review** of the subagent system before considering it done — 5 slices, one focused review per slice, every non-clean finding reproduced to a `file:line`. **Zero reachable production bugs;** this closes the edge-case contract violations and the test gaps.

> Note: the misleadingly-narrow original commit message is an artifact of how this got committed — the actual changeset is everything below (verified: 9 files, +202/−4).

## Fixes
- **`buildBoundedTranscript`** now honors `≤ maxChars` for **all** inputs (drops the elision marker, then hard-caps, when `maxChars` is below the marker reserve). Was only violated for `maxChars < 64` — never reached in production (`maxChars` is always `window*2` or the 48k fallback), but a real contract break, and the "bound can be exceeded" class flagged twice in review of #79.
- **`renderAgentTree`** renders `[]` below `MIN_TREE_COLUMNS` (8) rather than emit a line wider than `columns-1` (the fixed row prefix can't shrink below ~5 cols) — honors the stated no-self-wrap invariant at any width.
- **`resultPayload`** accepts `unknown` + `isRecord`-guards, so a malformed tool call (`undefined`/non-object args) can't throw; legacy `{ result }` and empty still handled.

## Tests (lock the fixes + previously-untested verified invariants)
- `buildBoundedTranscript` bound holds for `maxChars` 1..400 (fuzz)
- `renderAgentTree` ≤ `columns-1` across widths 1..40, and `[]` below the minimum
- `resultPayload` robustness (undefined / null / string / empty / legacy)
- `spawn_agent` is **never** advertised to a subagent (recursion capped at 1 — was uncovered)
- `makeLimiter` releases the slot when a body throws (no deadlock)
- spawn-abort returns an aborted result without resolving/running the model
- `OutputRouter` sink cleared even for an agent that produced no output (no leak)

## Docs
- Add the missing **`agent / subagents`** entry to `docs/harness-subsystems.md` (the manifest is part of the review contract; it had no subagent subsystem).

## Verified clean (with `file:line`)
Read-only enforcement (3 structural layers), no-recursion, forced-investigation, structured-output shapes, termination/status codes, policy classification + full mode matrix, spec precedence, semaphore correctness, sibling isolation, fresh-provider-per-child, lazy TsService, contextWindow threading, lifecycle ordering + not-echoed, no sink leak, kill-switch, stream reassembly.

Full `bun run validate`: **2049 pass / 0 fail**, all PTY suites green.

### CODE REVIEW

File: packages/core/src/agent/agent-runner.ts [hunk 1]
```diff
@@ -137,8 +137,18 @@ export function buildBoundedTranscript(
...
+  if (assembled.length <= maxChars) {
+    return assembled;
+  }
+
+  // Only reachable when maxChars < the marker reserve (never in production, where
+  // maxChars is window*2). Drop the marker, then hard-cap, so the ≤ maxChars
+  // contract holds for ALL inputs.
+  const bare = picked.join("\n\n");
+
+  return bare.length <= maxChars ? bare : bare.slice(0, maxChars);
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts]: Fixed (commit 568fb66). The hard-cap now checks whether the final code unit is a high surrogate (0xD800–0xDBFF) and drops it, so a split pair never leaves a lone surrogate. This path is only reachable for maxChars below the marker reserve (never in production), but the contract now holds cleanly.

File: packages/core/src/render/agent-tree.ts [hunk 2]
```diff
@@ -270,6 +277,11 @@ export function renderAgentTree(
   }
 
   const columns = opts.columns > 0 ? opts.columns : 80;
+
+  if (columns < MIN_TREE_COLUMNS) {
+    return [];
+  }
+
   const color = opts.color ?? true;
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/render/agent-tree.ts:285]: Good catch — real gap in the width fix. The overflow tail (`… +N more`) was pushed without clipping, so it self-wrapped at columns 8–14 where overflow can still occur. Fixed (42b25a6): the tail is now `fitLabel`-clipped to columns-1 like every row. Regression test asserts every line ≤ columns-1 across widths 8..30 WITH overflow (maxRows=4).

File: docs/harness-subsystems.md [hunk 3]
```diff
@@ -240,6 +240,51 @@ detectors, incl. function/brace-group bodies), `tests/policy-config.test.ts` (ru
...
+  not fatal. `TSFORGE_NO_DELEGATION=1` withholds delegation entirely (single-stream).
+
+**Risk areas** a mutation or `spawn_agent` slipping past the read-only filter (recursion); a
+compaction path that overflows or infinite-loops; a limiter slot leaked on throw (deadlock);
+a sink leak mis-routing a later turn's output; the turn-loop spawn-batch reordering edits
+after a spawn.
+
+**Checklist** `tests/agent-runner.test.ts`, `tests/agent-compaction.test.ts`,
+`tests/agent-structured-output.test.ts`, `tests/agent-scheduler.test.ts`,
+`tests/spawn-concurrency.test.ts`, `tests/spawn-ordering-invariant.test.ts`,
```
[the remark this answers is not in the document]
Comment (agjs) [reply docs/harness-subsystems.md]: Fixed (42b25a6). `tests/spawn-ordering-invariant.test.ts` never existed — the spawn-batch + edit-before-spawn ordering coverage is in `tests/tool-accounting.test.ts`, and the manifest checklist now points there.
